runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/FanOutStreamingEngineWorkerHarness.java

```diff
@@ -271,7 +280,10 @@ private void consumeWorkerMetadata(WindmillEndpoints windmillEndpoints) {
     synchronized (metadataLock) {
       // Only process versions greater than what we currently have to prevent double processing of
       // metadata. workerMetadataConsumer is single-threaded so we maintain ordering.
-      if (windmillEndpoints.version() > pendingMetadataVersion) {
+      // But in case the endpoint type in worker metadata is different from the active endpoint
+      // type, also process those endpoints
+      if (windmillEndpoints.version() > pendingMetadataVersion
+          || windmillEndpoints.endpointType() != activeEndpointType) {
```

Comment:
I think this should only update if the endpoint changes and the version is equal to the previous version.  We don't want to move backwards from version 5 to version 3 for example even if active enpdoint type changes.